loglib/logreader.c

```diff
@@ -0,0 +1,226 @@
+/*
+* Author: Christian Huitema
+* Copyright (c) 2019, Private Octopus, Inc.
+* All rights reserved.
+*
+* Permission to use, copy, modify, and distribute this software for any
+* purpose with or without fee is hereby granted, provided that the above
+* copyright notice and this permission notice appear in all copies.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+* ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL Private Octopus, Inc. BE LIABLE FOR ANY
+* DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+* ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include <stdio.h>
+#include <string.h>
+#include <inttypes.h>
+#include "picoquic_internal.h"
+#include "bytestream.h"
+#include "logreader.h"
+#include "logwriter.h"
+#include "cidset.h"
+
+static int byteread_packet_header(bytestream * s, picoquic_packet_header * ph);
+
+int fileread_binlog(FILE* bin_log, int(*cb)(bytestream*, void*), void* cbptr)
+{
+    int ret = 0;
+    uint8_t head[4];
+    bytestream_buf stream_msg;
+
+    fseek(bin_log, 16, SEEK_SET);
+
+    while (ret == 0 && fread(head, sizeof(head), 1, bin_log) > 0) {
+
+        uint32_t len = (head[0] << 24) | (head[1] << 16) | (head[2] << 8) | head[3];
+        if (len > sizeof(stream_msg.buf)) {
+            ret = -1;
+        }
+
+        if (ret == 0 && fread(stream_msg.buf, len, 1, bin_log) <= 0) {
+            ret = -1;
+        }
+
+        if (ret == 0) {
+            bytestream* s = bytestream_buf_init(&stream_msg, len);
+            ret |= cb(s, cbptr);
+        }
+    }
+
+    return ret;
+}
+
+typedef struct convert_log_file_event_st {
+
+    const picoquic_connection_id_t * cid;
+    binlog_convert_cb_t * callbacks;
+
+} convert_log_file_event_t;
+
+static int binlog_convert_event(bytestream * s, void * ptr)
+{
+    int ret = 0;
+    convert_log_file_event_t* ctx = (convert_log_file_event_t*)ptr;
+    void * cbptr = ctx->callbacks->ptr;
+
+    picoquic_connection_id_t cid;
+    ret |= byteread_cid(s, &cid);
+
+    /* filter for connection id */
+    if (ret != 0 || picoquic_compare_connection_id(&cid, ctx->cid) != 0) {
+        return ret;
+    }
```

Comment:
Is the CID test specific to this particular callback?

Reply:
Parsing the log file for one specific connection is the main use case for the log file event enumeration. Therefore it's done early so you can skip all the unnecessary parsing in case the log event doesn't belong to the connection you are looking for.